paddle/fluid/memory/allocation/allocator_facade.cc

```diff
@@ -223,25 +248,144 @@ class AllocatorFacadePrivate {
-#ifdef PADDLE_WITH_XPU
-    int device_count = platform::GetXPUDeviceCount();
-    for (int i = 0; i < device_count; ++i) {
-      platform::XPUPlace p(i);
-      system_allocators_[p] = std::make_shared<NaiveBestFitAllocator>(p);
+#if defined(PADDLE_WITH_CUDA) || defined(PADDLE_WITH_HIP)
+  const std::shared_ptr<Allocator>& GetAllocator(
+      const platform::CUDAPlace& place, const gpuStream_t& stream) {
+    auto place_it = cuda_allocators_.find(place);
+    PADDLE_ENFORCE_NE(place_it, cuda_allocators_.end(),
+                      platform::errors::NotFound(
+                          "No allocator found for the place %s", place));
+
+    const std::map<gpuStream_t, std::shared_ptr<Allocator>>& allocator_map =
+        place_it->second;
+    auto stream_it = allocator_map.find(stream);
+    PADDLE_ENFORCE_NE(
+        stream_it, allocator_map.end(),
+        platform::errors::NotFound(
+            "No allocator found for stream %s in place %s", stream, place));
+
+    return stream_it->second;
+  }
+
+  gpuStream_t GetDefaultStream() { return default_stream_; }
+
+  void RecordStream(Allocation* allocation, const gpuStream_t& stream) {
+    PADDLE_ENFORCE_EQ(
+        platform::is_gpu_place(allocation->place()), true,
+        platform::errors::InvalidArgument(
+            "Not allow to record stream for an allocation with place %s",
+            allocation->place()));
+    dynamic_cast<StreamSafeCUDAAllocation*>(allocation)->RecordStream(stream);
+  }
+
+  AllocationPtr CUDAAlloc(const platform::CUDAPlace& place,
+                          const gpuStream_t& stream, size_t size) {
+    std::shared_ptr<Allocator> cuda_allocator;
+    try {
+      cuda_allocator = GetAllocator(place, stream);
+    } catch (platform::EnforceNotMet& err) {
+      VLOG(9) << "No allocator found for stream " << stream << " in place "
+              << place << " , build a new one";
+      std::unique_lock<std::mutex> lock(cuda_init_mutex_);
+      try {
+        cuda_allocator = GetAllocator(place, stream);
+        VLOG(9) << "Other thread had build a new allocator for stream "
+                << stream << " in place " << place;
+      } catch (platform::EnforceNotMet& err) {
+        InitStreamSafeCUDAAllocator(place, stream);
+        cuda_allocator = GetAllocator(place, stream);
+      } catch (...) {
+        throw;
+      }
+    } catch (...) {
+      throw;
     }
+
+    try {
+      return cuda_allocator->Allocate(size);
+    } catch (BadAlloc&) {
+      VLOG(9) << "Allocation failed when allocating " << size
+              << " bytes for stream " << stream;
+      for (auto pair : cuda_allocators_[place]) {
+        pair.second->Release(place);
+      }
+      try {
+        return cuda_allocator->Allocate(size);
+      } catch (...) {
+        VLOG(9) << "Still allocation failed "
+                << "after release memory from all streams";
+        throw;
+      }
+    } catch (...) {
+      throw;
```

Comment:
Here can be removed to StreamSafeCUDAAllocator

Reply:
done, thx